reducer/src/test/java/com/graphicsfuzz/reducer/reductionopportunities/LiteralToUniformReductionOpportunitiesTest.java

```diff
@@ -25,40 +26,277 @@
+    CompareAsts.assertEqualAsts(vertexShaderReplaced, shaderJob.getVertexShader().get());
+
+    assertTrue(shaderJob.getPipelineInfo().hasUniform(Constants.FLOAT_LITERAL_UNIFORM_VALUES));
+  }
+
+  @Test
+  public void testReplaceMultipleLiterals() throws Exception {
+
+    final String vertexShader = "void main() { "
+        + "float a = 1.0;"
+        + "float b = 2.0;"
+        + "int c = 1;"
+        + "int d = 2;"
+        + "int e = 3;"
+        + "}";
+
+    final String vertexShaderReplaced =
+        "uniform int _GLF_uniform_int_values[3];"
+        + "uniform float _GLF_uniform_float_values[3];"
+        + "void main()"
+        + "{"
+        + "  float a = _GLF_uniform_float_values[0];"
+        + "  float b = _GLF_uniform_float_values[1];"
+        + "  int c = _GLF_uniform_int_values[0];"
+        + "  int d = _GLF_uniform_int_values[1];"
+        + "  int e = _GLF_uniform_int_values[2];"
+        + "}";
+
+    final String fragmentShader = "void main() { "
+        + "float a = 3.0;"
+        + "float b = 2.0;"
+        + "float c = 1.0;"
+        + "int d = 1;"
+        + "int e = 2;"
+        + "}";
+
+    final String fragmentShaderReplaced =
+          "uniform int _GLF_uniform_int_values[3];"
+        + "uniform float _GLF_uniform_float_values[3];"
+        + "void main()"
+        + "{"
+        + "  float a = _GLF_uniform_float_values[2];"
+        + "  float b = _GLF_uniform_float_values[1];"
+        + "  float c = _GLF_uniform_float_values[0];"
+        + "  int d = _GLF_uniform_int_values[0];"
+        + "  int e = _GLF_uniform_int_values[1];"
+        + "}";
+
+    final List<TranslationUnit> shaders = new ArrayList<>();
+    shaders.add(ParseHelper.parse(vertexShader, ShaderKind.VERTEX));
+    shaders.add(ParseHelper.parse(fragmentShader, ShaderKind.FRAGMENT));
+
+    final PipelineInfo pipelineInfo = new PipelineInfo();
+    final ShaderJob shaderJob = new GlslShaderJob(Optional.empty(),
+        pipelineInfo, shaders);
+    assertEquals(0, pipelineInfo.getNumUniforms());
+
+    final List<LiteralToUniformReductionOpportunity> ops =
+        LiteralToUniformReductionOpportunities
+            .findOpportunities(shaderJob,
+                new ReducerContext(false, ShadingLanguageVersion.ESSL_100, new RandomWrapper(0),
+                    new IdGenerator()));
+
+    assertEquals("There should be 10 opportunities", 10, ops.size());
+    assertEquals(0, pipelineInfo.getNumUniforms());
+
+    ops.forEach(AbstractReductionOpportunity::applyReduction);
+
+    CompareAsts.assertEqualAsts(fragmentShaderReplaced, shaderJob.getFragmentShader().get());
+    CompareAsts.assertEqualAsts(vertexShaderReplaced, shaderJob.getVertexShader().get());
+
+    assertTrue(shaderJob.getPipelineInfo().hasUniform(Constants.FLOAT_LITERAL_UNIFORM_VALUES));
+    assertTrue(shaderJob.getPipelineInfo().hasUniform(Constants.INT_LITERAL_UNIFORM_VALUES));
+  }
```

Comment:
I suggest adding a test where you apply just one reduction opportunity at first. Then check that just the one shader had the array uniform added, and the other shader stays the same (it does not have the array uniform). Then manually add another uniform to the pipeline info. Then find and apply the remaining opportunities in both shaders. Check that the bindings for the manually added uniform and the existing array uniform have not changed. I suggest this because I think this test (or a similar test) would currently fail.

Reply:
A new test case added accordingly to the suggestion.